File: javascript/138.copy-list-with-random-pointer.js
/*
 * @lc app=leetcode id=138 lang=javascript
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/**
 * // Definition for a Node.
 * function Node(val, next, random) {
 *    this.val = val;
 *    this.next = next;
 *    this.random = random;
 * };
 */

/**
 * @param {Node} head
 * @return {Node}
 */
var copyRandomList = function(head) {
  // Build hashmap of nodes, then build copy
  // O(n) time
  // O(n) space
  if (!head) return null
  let nodeMap = new Map()
  let currNode = head
  // create map of nodes
  while (currNode) {
    nodeMap.set(currNode, new Node(currNode.val))
    currNode = currNode.next
  }
  // build copy with pointers
  currNode = head
  while (currNode) {
    nodeMap.get(currNode).next = nodeMap.get(currNode.next) ?? null
    nodeMap.get(currNode).random = nodeMap.get(currNode.random) ?? null
    currNode = currNode.next
  }
  return nodeMap.get(head)
};
// @lc code=end

File: cpp/138.copy-list-with-random-pointer.cpp
/*
 * @lc app=leetcode id=138 lang=cpp
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution
{
public:
    Node *copyRandomList(Node *head)
    {
        // construct a hashmap of nodes, then build deep copy
        // O(n) time O(n) space
        if (!head)
            return head;
        unordered_map<Node *, Node *> nodeMap;
        Node *currNode = head;
        // create hashmap of nodes
        while (currNode)
        {
            nodeMap[currNode] = new Node{currNode->val};
            currNode = currNode->next;
        }
        // build deep copy with pointers
        currNode = head;
        while (currNode)
        {
            nodeMap[currNode]->next = nodeMap[currNode->next];
            nodeMap[currNode]->random = nodeMap[currNode->random];
            currNode = currNode->next;
        }
        return nodeMap[head];
    }
};
// @lc code=end
